### ISSUE TITLE
systemd-bless-boot integration into user session

### ISSUE THREAD

**Issue by AdrianVovk** [systemd/systemd#35019]
### Is your feature request related to a problem? Please describe

The bless-boot architecture is currently restrained to the system service manager. However, on user systems its often better to also make sure that user sessions are stable and functional before disabling bootcounting.

For example: a system where GDM is broken will still _boot_, but the user cannot log in and thus the system is unusable.

### Describe the solution you'd like

We introduce a new `systemd-check-user-session.service` as part of `boot-complete.target`. If enabled, it simply creates a socket in /run and then waits for something to send a message on there. Once that happens, it exits, allowing `boot-complete.target` to be reached, and allowing `systemd-bless-boot.service` to run.

Then, we can duplicate `boot-complete.target` and the subsequent `systemd-bless-boot.service` in the user manager. The user's instance of `systemd-bless-boot.service` just connects to the socket and tells `systemd-check-user-session.service` to exit. These units are conditionally started only if the socket in /run exists, so if the mechanism is disabled on a system level that propagates all the way through to the user session.

Desktop environments can then trivially drop whatever checks they want into the user session, and make them part of the user session's `boot-complete.target`.

### Describe alternatives you've considered

The desktop environments can do this themselves. But then there will be bespoke per-DE solutions for this, which means DEs will be the only ones in charge of the determination. Doing it in systemd ensures that the mechanism is easily extensible with custom logic

### The systemd version you checked that didn't have the feature you are asking for

257

**Comment by poettering** [comment]
Yeah, sounds OK. But please make sure this service can be instantiated, so that multiple userspace components can instantiate it independently. for example gdm would pull in systemd-check-user-session@gdm.service and it would bind /run/systemd/check-user-session/gdm or so, which gdm would then talk to. 

And the service should be talking varlink I guess, so that we can maybe pass more metadata along. And it probably should do PK, so that we can control that only specific clients can unlock the blessing.

**Comment by poettering** [comment]
btw, this came up before, @jwrdegoede iirc wanted this. No idea what happened the last time this was discussed though. @jwrdegoede can you fill us in?

**Comment by jwrdegoede** [comment]
> btw, this came up before, @jwrdegoede iirc wanted this. No idea what happened the last time this was discussed though. @jwrdegoede can you fill us in?

Last time we discussed this your suggestion was to add varlink support to the systemd-bless-boot command and allow running it as a varlink service. And then have the desktop-environment use varlink to bless the boot for user session based boot blessing.

I like the idea to integrate this into the user services, although I wonder how that will work for desktop-environments which lack systemd user-session integration.

I have not done any work on this, so if @AdrianVovk wants to go ahead and implement this I'm very much in favor of that.

**Comment by AdrianVovk** [comment]
> Last time we discussed this your suggestion was to add varlink support to the systemd-bless-boot command and allow running it as a varlink service. And then have the desktop-environment use varlink to bless the boot for user session based boot blessing.

I think this could be useful for the opposite reason, allowing the desktop environment to un-bless the boot (ie mark it as bad). If something critically bad happens, we can completely bypass the boot counting and go straight to disabling the version

> I like the idea to integrate this into the user services, although I wonder how that will work for desktop-environments which lack systemd user-session integration.

Should be fine. The user manager is running even if the DE doesn't use it to manage itself

**Comment by AdrianVovk** [comment]
(In https://github.com/systemd/systemd/issues/36844 there is a specific proposal for a varlink API)
